solidity/contracts/MultiAssetVault.sol

```diff
@@ -0,0 +1,240 @@
+            revert AssetNotSupported();
+        }
+        if (amount == 0) {
+            revert InvalidAmount(amount);
+        }
+        if (depositOwner == address(0)) {
+            revert ZeroAddress();
+        }
+        if (depositOwner == asset) {
+            revert DepositOwnerCannotBeAsset();
+        }
+
+        // Transfer the asset from the deposit owner to the vault and approve
+        // Mezo Portal to pull them.
+        IERC20(asset).safeTransferFrom(msg.sender, address(this), amount);
+        IERC20(asset).forceApprove(address(mezoPortal), amount);
+
+        // Create a deposit in Mezo Portal.
+        // 0 denotes no lock period for this deposit.
+        mezoPortal.deposit(address(asset), uint96(amount), 0);
+
+        // MezoPortal doesn't return depositId, so we have to read depositCounter
+        // which assigns depositId to the current deposit.
+        uint256 depositId = mezoPortal.depositCount();
+
+        // slither-disable-next-line reentrancy-events
+        emit DepositCreated(depositOwner, asset, depositId, amount);
+
+        // Register the deposit info.
+        // slither-disable-next-line reentrancy-benign
+        deposits[depositOwner][asset][depositId] = DepositInfo({
+            balance: amount
+        });
+
+        return depositId;
+    }
+
+    /// @notice Withdraws all deposited assets from the given deposit.
+    /// @dev It can be called by the deposit owner only.
+    /// @param asset Address of the asset to be withdrawn.
+    /// @param depositId ID of the deposit.
+    /// @param receiver Address of the account to which the assets will be sent.
+    /// @return The amount of assets withdrawn.
+    function withdraw(
+        address asset,
+        uint256 depositId,
+        address receiver
+    ) external returns (uint256) {
+        if (receiver == address(0)) {
+            revert ZeroAddress();
+        }
+        if (receiver == asset) {
+            revert ReceiverCannotBeAsset();
+        }
+
+        uint256 depositedAmount = deposits[msg.sender][asset][depositId]
+            .balance;
+
+        // Check if the deposit exists. If the deposit was not created or was
+        // already withdrawn, the depositedAmount will be 0.
+        if (depositedAmount == 0) {
+            revert DepositNotFound();
+        }
+
+        // Delete the deposit info, as the deposit is being fully withdrawn.
+        delete deposits[msg.sender][asset][depositId];
+
+        // Withdraw the assets from the MezoPortal. Check if the withdrawn amount
+        // matches the deposited amount.
+        uint256 initialBalance = IERC20(asset).balanceOf(address(this));
+
+        mezoPortal.withdraw(address(asset), depositId);
+
+        uint256 withdrawnAmount = IERC20(asset).balanceOf(address(this)) -
+            initialBalance;
+
+        if (withdrawnAmount != depositedAmount) {
+            revert UnexpectedWithdrawnAmount(withdrawnAmount, depositedAmount);
+        }
```

Comment:
Just a heads up this is currently possible only in two scenarios that are not triggerable for `MultiAssetVault` contract: 
- WBTC -> tBTC migration in the Mezo Portal when decimals are switched.
- Deposit receipt (stBTC) minted and repaid for non-zero fees.

Reply:
Yes, I'm aware that is not possible currently. I just wanted to protect us from unexpected upgrades or a bug in the Mezo Portal. With this check, we won't delete the deposit if something odd happens.

Reply:
Yep, this check makes sense. I only wanted to make sure we are in sync about the scenarios possible to trigger it.